theano/gpuarray/ctc.py

```diff
@@ -0,0 +1,185 @@
+                             context_name=context)()
+        outputs = [costs]
+
+        if self.compute_grad:
+            gradients = GpuArrayType(dtype='float32',
+                                     broadcastable=(False, False, False,),
+                                     context_name=context)()
+            outputs += [gradients]
+
+        return theano.Apply(self, inputs=[t_activations, t_labels, t_input_lengths],
+                            outputs=outputs)
+
+    def L_op(self, inputs, outputs, output_grads):
+        if not ctc_enabled:
+            raise RuntimeError('Baidu CTC is not enabled and '
+                               'GpuConnectionistTemporalClassification Op '
+                               'can not be constructed.')
+        # Gradients computed by Op
+        gradients = outputs[1]
+        # Gradients of original function, to compose chain rule
+        grad_op = output_grads[0]
+        grad_shuffle = GpuDimShuffle(input_broadcastable=(False, False, False,),
+                                     new_order=(1, 0, 2))(gradients)
+        grad_bdot = T.basic.batched_dot(grad_op, grad_shuffle)
+        grad_shuffle_reverse = GpuDimShuffle(input_broadcastable=(False, False, False,),
+                                             new_order=(1, 0, 2))(grad_bdot)
+        return [grad_shuffle_reverse,
+                grad_undefined(self, 1, inputs[1]),
+                grad_undefined(self, 2, inputs[2])]
+
+
+def gpu_ctc(activations, labels, input_lengths):
+    """
+    Compute CTC loss function on the GPU.
+
+    Parameters
+    ----------
+    activations
+        Three-dimensional tensor, which has a shape of (t, m, p), where
+        t is the time index, m is the minibatch index, and p is the index
+        over the probabilities of each symbol in the alphabet. The memory
+        layout is assumed to be in C-order, which consists in the slowest
+        to the fastest changing dimension, from left to right. In this case,
+        p is the fastest changing dimension.
+    labels
+        A 1-D tensor of all the labels for the minibatch.
+    input_lengths
+        A 1-D tensor with the number of time steps for each sequence in
+        the minibatch.
+
+    Returns
+    -------
+    1-D array
+        Cost of each example in the minibatch.
+    """
+    return GpuConnectionistTemporalClassification()(activations, labels,
+                                                    input_lengths)
+
+
+# Disable gradient computation if not needed
+@register_canonicalize
+@register_stabilize
+@local_optimizer([GpuConnectionistTemporalClassification])
+def local_gpu_ctc_no_grad(node):
+    if isinstance(node.op, GpuConnectionistTemporalClassification):
+        if len(node.outputs) > 1:
+            if len(node.outputs[1].clients) == 0:   # gradient is not used
+                node.op = GpuConnectionistTemporalClassification(compute_grad=False)
+                node.outputs = node.outputs[:1]   # costs only
```

Comment:
You should NEVER modify a node inplace.  Optimizations are supposed to return new outputs to replace the ones of the current node.